r/R/trackplots.R

```diff
@@ -180,14 +180,22 @@ trackplot_create_arrow_segs <- function(data, region, size = 50, head_only = FAL
         new_arrow,
         data %>% dplyr::select(-c("start", "end")) %>% dplyr::slice(i)
       )))
+    } else {
+      arrow_list <- c(arrow_list, list(dplyr::slice(data, i)))
     }
   }
   arrows <- dplyr::bind_rows(arrow_list)
   if (head_only) {
     # Set segment size small enough to be invisible if head_only
+    # Cannot swap values with if-else mutation block due to sequential evaluation
+    arrows_shortened <- dplyr::mutate(
+      arrows,
+      start = ifelse(strand, arrows$end - 1e-4, arrows$start),
+      end = ifelse(strand, arrows$end, arrows$start+1e-4)
+    )
     arrows <- arrows %>% dplyr::mutate(
-      start = ifelse(strand, end - 1e-4, start),
-      end = ifelse(strand, end, start + 1e-4)
+      start = arrows_shortened$start,
+      end = arrows_shortened$end
     )
```

Comment:
On a third look at this, there is no bug in the original -- the only time the `end = ` case reads the `start` variable is if `strand` is `FALSE`, but in that case `start` hasn't been modified. I discovered this when trying to add tests for the `head_only` case and confirming the previous code was actually broken.

That said, for an approach that would be less dependent on the specific logic in this case I found a weird-looking alternative. It is still not great but a tiny bit simpler:

```r
arrows <- arrows %>% dplyr::mutate(
    start_tmp = start,
    start = ifelse(strand, end-1e-4, start),
    end = ifelse(strand, end, start_tmp+1e-4),
    start_tmp = NULL
)
```

Confirmed to work on a test tibble `arrows <- tibble(start=1:3, end=11:13, strand=c(TRUE, FALSE, TRUE))`